### PR TITLE
Workspace termination time metrics

### PR DESCRIPTION
**Pull request by sparkoo**
### What does this PR do?
Adds new metrics measuring time of workspace termination.

workspace_stoptime

### What issues does this PR fix or reference?
#13500

#### Release Notes
n/a

#### Docs PR
https://github.com/eclipse/che-docs/pull/747

### CODE REVIEW

**Comment by musienko-maxim** [comment]
Can one of the admins verify this PR?

**Comment by sleshchenko** [comment]
ci-build

**Review by skabashnyuk — COMMENTED** [review]
Can you make a screenshot of how panels look like?

**Comment by sparkoo** [comment]
screenshot

**Comment by skabashnyuk** [comment]
On the dashboard, we are using two new metrics. I wonder are we adding 3 new metrics (max, avg, count) or 3 + distribution(for heat map). Since we are not adding heat map as we do that for workspace start, can you make sure we are not adding unused metric?

File: wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WorkspaceStopTrackerMeterBinder.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            .minimumExpectedValue(Duration.ofSeconds(10))
+            .maximumExpectedValue(Duration.ofMinutes(15))
+            .register(registry);
+
+    // only subscribe to the event once we have the counters ready
+    eventService.subscribe(this::handleWorkspaceStatusChange, WorkspaceStatusEvent.class);
+  }
+
+  private void handleWorkspaceStatusChange(WorkspaceStatusEvent event) {
+    if (event.getPrevStatus() == WorkspaceStatus.RUNNING
```
Comment (metlos) [inline wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WorkspaceStopTrackerMeterBinder.java]: I think the `STARTING` -> `STOPPING` transition is also valid. In the case of that transition we wouldn't record the start of the stop time and then issue an unnecessary warning once the workspace is stopped.

Comment (sparkoo) [reply wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WorkspaceStopTrackerMeterBinder.java]: I've followed the ones that are commented as `normal behaviour` here https://github.com/eclipse/che/blob/master/core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/workspace/WorkspaceStatus.java
But `STARTING` -> `STOPPING` also looks like a valid transition. I'm not sure. @skabashnyuk ?

**Comment by sparkoo** [comment]
> On the dashboard, we are using two new metrics. I wonder are we adding 3 new metrics (max, avg, count) or 3 + distribution(for heat map). Since we are not adding heat map as we do that for workspace start, can you make sure we are not adding unused metric?

updated
```
/ $ curl http://localhost:8087 | grep ^che_workspace_stop_time
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 29105  100 29105    0     0  2186k      0 --:--:-- --:--:-- --:--:-- 2186k
che_workspace_stop_time_seconds_count{area="workspace",result="success",} 3.0
che_workspace_stop_time_seconds_sum{area="workspace",result="success",} 2.669
che_workspace_stop_time_seconds_max{area="workspace",result="success",} 0.464
```

File: wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WorkspaceStopTrackerMeterBinder.java (hunk 1 above)
Comment (metlos) [reply wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WorkspaceStopTrackerMeterBinder.java]: https://github.com/eclipse/che/blob/master/core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/workspace/WorkspaceStatus.java#L60 ;)

Comment (skabashnyuk) [reply wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WorkspaceStopTrackerMeterBinder.java]: STARTING -> STOPPING - yes. This the case of start interruption

**Comment by metlos** [comment]
ci-build

**Comment by sparkoo** [comment]
ci-build

**Review by skabashnyuk — APPROVED** [review]

**Comment by skabashnyuk** [comment]
ci-test

**Comment by metlos** [comment]
ci-build

**Review by metlos — APPROVED** [review]

**Comment by sparkoo** [comment]
updated image and added docs PR (https://github.com/eclipse/che-docs/pull/747)

**Comment by skabashnyuk** [comment]
CC @ibuziuk

**Pull request merged**